### PR TITLE
Feat/api hardening

### PR DESCRIPTION
**Pull request by jshessen**
# PR: feat/api-hardening → main

## Summary

Delivers three layers of hardening on top of the v0.3.24 dependency modernization:

1. **Pydantic v2 migration** — converts `Series` date parsing from a manual `validator`
   to a proper `field_validator`, eliminating the `@validator` deprecation warning.
2. **Configurable timeouts & safer logging** — exposes `read_timeout` / `connect_timeout`
   on the client, replaces the linear retry strategy with `tenacity` exponential backoff,
   and redacts credentials from log output.
3. **API hardening** — adds input validation on all public `MeterReader` parameters,
   propagates `aggregation_level` and `units` query params to the API, and adds null-response
   guards throughout.

Also fixes an upstream defect from PR #48: `low_read_limit` on `SignalStrength` was
declared as a required `Optional[int]` field (missing `= None`), causing `ValidationError`
on real API responses.

> **Chain note:** The original plan split this into three separate branches
> (`feat/pydantic-v2-migration`, `feat/configurable-timeouts`, `feat/api-hardening`).
> After rebasing onto the updated `main` (post PR #48 merge), those three were folded
> into a single stack. `feat/configurable-timeouts` remains as a branch alias pointing
> to the same tip (`7f59f13`).

## Target branch

`main` (base: `0985ea8` — PR #48: chore/bump-dependencies, v0.3.24)

## Commits

| SHA | Message |
| --- | --- |
| `c5d74b4` | chore: lint fixes for core library and existing tests |
| `7cbef67` | feat: migrate Series date parsing to pydantic v2 field_validator |
| `d93e57b` | chore: lint fixes for pydantic-v2-migration files |
| `58eec29` | feat: add configurable timeouts, exponential backoff, and safer logging |
| `5c967ae` | chore: lint fixes for configurable-timeouts tests |
| `c29e0d6` | feat: api hardening — input validation, aggregation/units params, null response guards |
| `0384641` | chore: lint fixes for api-hardening files |
| `7f59f13` | fix: low_read_limit Optional default, test_models_extensions unit casing; v0.3.25 |

## Files changed (16 files, +1007 −207)

| File | Change |
| --- | --- |
| `.flake8` | Suppress W503 (line-continuation before binary operator) |
| `pyonwater/account.py` | Pydantic v2 model updates; timeout kwargs forwarded to client |
| `pyonwater/client.py` | `read_timeout` / `connect_timeout` params; `tenacity` exponential backoff; redacted auth logging |
| `pyonwater/meter_reader.py` | `aggregation_level` / `units` query params; null-response guards; `ValidationError` propagation |
| `pyonwater/models/eow_historical_models.py` | Pydantic v2 `field_validator` for series timestamps |
| `pyonwater/models/eow_models.py` | `low_read_limit: Optional[int] = None` default (upstream defect fix) |
| `pyonwater/models/units.py` | Unit enum consolidation; case-sensitive validation enforced |
| `pyproject.toml` | Bump to `0.3.25`; add `tenacity` dependency |
| `tests/conftest.py` | Shared timeout fixtures; extended mock client helpers |
| `tests/test_api_contracts.py` | **New** — contract tests for all API response pydantic models |
| `tests/test_client.py` | Timeout, backoff, and credential-redaction coverage |
| `tests/test_meter.py` | Minor lint fixes (`@pytest.mark.asyncio`, `-> None`) |
| `tests/test_meter_reader.py` | Null-response, `ValidationError`, and aggregation-level tests |
| `tests/test_models_extensions.py` | Unit casing fix: `"gal"` → `"GAL"` |
| `tests/test_series_date_parsing.py` | **New** — series `field_validator` parsing tests |
| `tests/test_validation.py` | **New** — input validation coverage for `MeterReader` |

## Notes

- Bumps version `0.3.24` → `0.3.25`.
- All tests pass; pre-commit (black, isort, ruff, mypy, codespell) clean.
- `tenacity` added as a direct dependency; retry logic removed from `client.py` manual loops.
- The `"GAL"` casing fix in `test_models_extensions.py` is paired with the `low_read_limit`
  fix — both are corrections to defects introduced in the upstream PR #48 codebase.

### CODE REVIEW

**Comment by GeorgeIoak** [comment]
Thanks for the improvements!

I’ve reviewed current `main` + this PR, and I’m intentionally holding off on opening another PR to avoid overlap/conflicts. The main remaining area I have locally is export-range ingestion (`/reports/export_initiate` + status polling + CSV parse) with focused tests, which appears out of scope for this PR.

I’ll wait for this to land (and for any follow-up you may have) before proposing anything and then rebase on latest `main` to keep changes narrow and non-competitive.

I've been my running my version for weeks and haven't seen any issues but it's of course limited to what my eyeonwater account exposes and what I discovered.

**Comment by jshessen** [comment]
> Thanks for the improvements!
> 
> I’ve reviewed current `main` + this PR, and I’m intentionally holding off on opening another PR to avoid overlap/conflicts. The main remaining area I have locally is export-range ingestion (`/reports/export_initiate` + status polling + CSV parse) with focused tests, which appears out of scope for this PR.
> 
> I’ll wait for this to land (and for any follow-up you may have) before proposing anything and then rebase on latest `main` to keep changes narrow and non-competitive.
> 
> I've been my running my version for weeks and haven't seen any issues but it's of course limited to what my eyeonwater account exposes and what I discovered.

My intent was to handle all interest into Eyeonwater using a single entity which could accurately handle the inconsistent timing of data from pyonwater and accurately calculate costs.

The changes here are primarily focused on exposing a few elements, but adding tests and type handling.

I have a few more branches, but I'll wait to submit those until this is cleared and your submission gets into the queue.

**Comment by GeorgeIoak** [comment]
My next PR is focused on the historical-import path, 
use the export flow (/reports/export_initiate -> /reports/export_check_status/{task_id} -> CSV parse) for bulk history loading, instead of the current day-by-day consumption calls in read_historical_data.

File: pyonwater/meter_reader.py [hunk 1]
```diff
@@ -53,34 +71,67 @@ async def read_meter_info(self, client: Client) -> MeterInfo:
         return meter_info
 
     async def read_historical_data(
-        self, client: Client, days_to_load: int
+        self,
+        client: Client,
+        days_to_load: int,
+        aggregation: AggregationLevel = AggregationLevel.HOURLY,
```
Comment (kdeyev) [inline pyonwater/meter_reader.py:77]: **Medium**: `asyncio.get_event_loop()` is deprecated since Python 3.10 and emits a `DeprecationWarning` when there is no running loop. Since this code always runs inside an async context, consider replacing with `asyncio.get_running_loop()` which is the recommended alternative.

File: pyonwater/models/units.py [hunk 2]
```diff
@@ -29,3 +29,31 @@ class NativeUnits(str, Enum):
...
     CF = "cf"
     CM = "cm"
+
+
+class RequestUnits(str, Enum):
+    """Enum of units for API requests (different from response units)."""
+
+    GALLONS = "gallons"
+    CUBIC_FEET = "cf"
+    CCF = "ccf"
```
Comment (kdeyev) [inline pyonwater/models/units.py:39]: **Low / Naming**: `QUARTER_HOURLY = "hr"` — the wire value `"hr"` reads like "hourly", which is confusing for someone encountering this enum for the first time. A short doc-comment on the enum member (or a class docstring mapping) would help. If `"hr"` truly means 15-minute intervals on the EyeOnWater side, a clarifying comment inline would be appreciated beyond the trailing comment.

**Review by kdeyev — COMMENTED** [review]
## PR #50 Review — feat/api-hardening

Overall this is a solid hardening pass: timeouts, retries, input validation, and flexible date parsing are all welcome additions. A few items to consider:

**Additional nits not attached inline:**

1. **`tests/test_validation.py`** — Some tests (e.g., `test_empty_meter_uuid_raises_value_error`, `test_negative_days_raises_value_error`) are testing Python/pydantic semantics rather than library-specific logic. They're not harmful but add maintenance surface without covering real edge cases.

2. **Pervasive `# nosec: B101`** — Almost every `assert` in tests is annotated with `# nosec: B101`. Consider adding a per-file bandit ignore (e.g., `# nosec` at the top or a `.bandit` config with `skips: [B101]` for the tests directory) to reduce noise.

3. **`eow_historical_models.py` date validator** — The `"%Y"` format will match *any* 4-digit string (e.g., `"9999"` or `"0000"`). If a range check matters, consider bounding the year.

Good work on the `low_read_limit` bugfix and the tenacity-based retry logic! 👍

File: pyproject.toml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "pyonwater"
-version = "0.3.24"
+version = "0.3.25"
 description = "EyeOnWater client library."
 authors = []
```
Comment (kdeyev) [inline pyproject.toml:5]: **Low**: `extension-pkg-whitelist` has been deprecated by pylint in favor of `extension-pkg-allow-list`. Consider updating to the modern option name.

File: tests/conftest.py [hunk 4]
```diff
@@ -17,99 +25,153 @@ def is_unit(string: str) -> bool:
 
 
 def replace_units(data: Any, new_unit: str) -> Any:
-    """Replace EOW units in JSON recursively."""
```
Comment (kdeyev) [inline tests/conftest.py:20]: **Low**: The `_aggregation` parameter is accepted but never used in the response construction. If the intent is to validate that the client sends the correct aggregation value, consider asserting it matches the expected value. Otherwise the leading underscore is fine but a comment explaining it's intentionally ignored would help.

File: tests/conftest.py [hunk 5]
```diff
@@ -17,99 +25,153 @@ def is_unit(string: str) -> bool:
...
 
 
-async def mock_signin_endpoint(request):
+async def mock_signin_endpoint(_request: web.Request) -> web.Response:
     """Sign in HTTP endpoint mock."""
     return web.Response(text="Hello, world", headers={"cookies": "key=val"})
 
 
-def mock_get_meters_endpoint(request):
-    """Fetch meters endpoit mock."""
```
Comment (kdeyev) [inline tests/conftest.py:41]: **Medium**: This references `tests/mock_data/at_a_glance_mock.json` which doesn't appear to exist in the repository or in this PR. Any test using `build_at_a_glance_endpoint` will fail with a `FileNotFoundError`. Is the mock data file missing from the PR?

File: pyonwater/meter_reader.py (hunk 1 above)
Comment (kdeyev) [inline pyonwater/meter_reader.py:77]: **Medium**: `asyncio.get_event_loop()` is deprecated since Python 3.10 and emits a `DeprecationWarning` when there is no running loop. Since this code always runs inside an async context, consider replacing with `asyncio.get_running_loop()` which is the recommended alternative.

File: pyonwater/models/units.py (hunk 2 above)
Comment (kdeyev) [inline pyonwater/models/units.py:39]: **Low / Naming**: `QUARTER_HOURLY = "hr"` — the wire value `"hr"` reads like "hourly", which is confusing for someone encountering this enum for the first time. A short doc-comment on the enum member (or a class docstring mapping) would help. If `"hr"` truly means 15-minute intervals on the EyeOnWater side, a clarifying comment inline would be appreciated beyond the trailing comment.

File: pyproject.toml (hunk 3 above)
Comment (kdeyev) [inline pyproject.toml:5]: **Low**: `extension-pkg-whitelist` has been deprecated by pylint in favor of `extension-pkg-allow-list`. Consider updating to the modern option name.

File: tests/conftest.py (hunk 4 above)
Comment (kdeyev) [inline tests/conftest.py:20]: **Low**: The `_aggregation` parameter is accepted but never used in the response construction. If the intent is to validate that the client sends the correct aggregation value, consider asserting it matches the expected value. Otherwise the leading underscore is fine but a comment explaining it's intentionally ignored would help.

File: tests/conftest.py (hunk 5 above)
Comment (kdeyev) [inline tests/conftest.py:41]: **Medium**: This references `tests/mock_data/at_a_glance_mock.json` which doesn't appear to exist in the repository or in this PR. Any test using `build_at_a_glance_endpoint` will fail with a `FileNotFoundError`. Is the mock data file missing from the PR?

**Review by kdeyev — COMMENTED** [review]
## PR #50 Review — feat/api-hardening

Overall this is a solid hardening pass: timeouts, retries, input validation, and flexible date parsing are all welcome additions. A few items to consider:

**Additional nits not attached inline:**

1. **`tests/test_validation.py`** — Some tests (e.g., `test_empty_meter_uuid_raises_value_error`, `test_negative_days_raises_value_error`) are testing Python/pydantic semantics rather than library-specific logic. They're not harmful but add maintenance surface without covering real edge cases.

2. **Pervasive `# nosec: B101`** — Almost every `assert` in tests is annotated with `# nosec: B101`. Consider adding a per-file bandit ignore (e.g., `# nosec` at the top or a `.bandit` config with `skips: [B101]` for the tests directory) to reduce noise.

3. **`eow_historical_models.py` date validator** — The `"%Y"` format will match *any* 4-digit string (e.g., `"9999"` or `"0000"`). If a range check matters, consider bounding the year.

Good work on the `low_read_limit` bugfix and the tenacity-based retry logic! 👍

**Comment by kdeyev** [comment]
This PR was reviewed with the assistance of agentic AI (GitHub Copilot, Claude Opus 4.6).

File: pyonwater/meter_reader.py (hunk 1 above)
Comment (jshessen) [reply pyonwater/meter_reader.py:77]: Fixed in 522f2a1 — replaced both occurrences of asyncio.get_event_loop() with asyncio.get_running_loop() in meter_reader.py.

File: pyonwater/models/units.py (hunk 2 above)
Comment (jshessen) [reply pyonwater/models/units.py:39]: Fixed in 522f2a1 — expanded the AggregationLevel class docstring and added an inline note on the QUARTER_HOURLY member explicitly calling out that the wire value "hr" represents 15-minute resolution despite the name suggesting hourly.

File: pyproject.toml (hunk 3 above)
Comment (jshessen) [reply pyproject.toml:5]: Fixed in 522f2a1 — updated to extension-pkg-allow-list = ["pydantic"] in [tool.pylint.MASTER].

File: tests/conftest.py (hunk 4 above)
Comment (jshessen) [reply tests/conftest.py:20]: Fixed in 522f2a1 — added a docstring to build_consumption_endpoint_with_aggregation explaining that _aggregation is intentionally accepted but not forwarded: the mock always returns the same fixture regardless of aggregation level. The leading underscore now has an explicit written justification.

**Review by jshessen — COMMENTED** [review]
Thanks for the review. All 5 items have been addressed in 522f2a1 ("fix: address PR #50 review feedback"):

asyncio.get_event_loop() deprecated — replaced both occurrences in meter_reader.py with asyncio.get_running_loop().

QUARTER_HOURLY = "hr" naming confusion — expanded the AggregationLevel class docstring and added an inline note on QUARTER_HOURLY explicitly calling out that the EyeOnWater wire value "hr" represents 15-minute resolution despite the name suggesting hourly.

extension-pkg-whitelist deprecated — updated pyproject.toml to extension-pkg-allow-list = ["pydantic"] in [tool.pylint.MASTER].

_aggregation unused parameter — added a docstring to build_consumption_endpoint_with_aggregation explaining that _aggregation is intentionally accepted but not forwarded: the mock always returns the same fixture regardless of aggregation level. The leading underscore now has an explicit written justification.

at_a_glance_mock.json missing — added at_a_glance_mock.json to the branch. The file existed in the downstream feat/at-a-glance branch but had been inadvertently omitted from this PR.
